src/pages/settings.js

```diff
@@ -0,0 +1,164 @@
+import React, { useState, useContext, useEffect } from "react"
+
+import Navbar from "../components/navbar"
+import Message from "../components/message"
+
+import { HeartwoodStateContext, HeartwoodDispatchContext } from "../state/HeartwoodContextProvider"
+import { isLoggedIn } from "../utils/utils"
+import { firestore } from "../firebase/firebase"
+
+const Settings = () => {
+  const dispatch = useContext(HeartwoodDispatchContext)
+  const state = useContext(HeartwoodStateContext)
+
+  const [form, setForm] = useState({firstName: "",
+    lastName: "",
+    userName: "",
+    dateOfBirth: "",
+    message: { text: "", type: "" }})
+
+  const onChangeHandler = (event) => {
+    const { name, value } = event.currentTarget
+
+    if (name === "userFirstName") {
+      setForm({ ...form, firstName: value })
+    } else if (name === "userLastName") {
+      setForm({ ...form, lastName: value })
+    } else if (name === "userName") {
+      setForm({ ...form, userName: value })
+    } else if (name === "dateOfBirth") {
+      setForm({ ...form, dateOfBirth: value })
+    }
+  }
+
+  // Read all inputs and update the user in session state and in the firestore
+  const updateProfile = () => {
+    // update the information in firestore
+    try {
+      firestore
+        .collection("users")
+        .doc(state.user.uid)
+        .update({firstName: form.firstName,
+          lastName: form.lastName,
+          userName: form.userName,
+          dateOfBirth: form.dateOfBirth})
+
+      // update the user that will be stored in state then save the user
+      const updatedUser = {...state.user,
+        firstName: form.firstName,
+        lastName: form.lastName,
+        userName: form.userName,
+        dateOfBirth: form.dateOfBirth}
+
+      dispatch({ type: "UPDATE", user: updatedUser })
+
+      // If successful tell the user
+      setForm({ ...form, message: { text: "Your preferences have been updated", type: "success" } })
+    } catch (error) {
+      setForm({ ...form, message: { text: "An error occured when updating preferences", type: "error" } })
+    }
+
+    // Hide the successfully sent notification after 3 seconds
+    setTimeout(() => {
+      setForm({ ...form, message: { text: "", type: "" } })
+    }, 3000)
+  }
+
+  useEffect(() => {
+    if (isLoggedIn(state.user)) {
+      // On load, set all the inputs to the user's current preferences
+      setForm({...form,
+        firstName: state.user.firstName,
+        lastName: state.user.lastName,
+        userName: state.user.userName,
+        dateOfBirth: state.user.dateOfBirth})
+    }
+  }, [state.user])
```

Comment:
Codacy found an issue: [React Hook useEffect has a missing dependency: 'form'. Either include it or remove the dependency array. You can also do a functional update 'setForm(f => ...)' if you only need 'form' in the 'setForm' call.](https://app.codacy.com/gh/treetoplearning/heartwood/pullRequest?prid=5919976)